### PR TITLE
Refactor Router Class to Separate Static and Instance Logic

### PR DESCRIPTION
**Pull request by adl5423**
Hello!

This pull request aims to refactor the existing Router class by separating its static and instance logic into two distinct classes: StaticRouter and Router. This separation enhances code maintainability, readability, and testability.

Changes:
- Created a new StaticRouter class that encapsulates all static methods and properties.
- Moved getPageHref and setPageHref from Router to StaticRouter.
- Modified the Router class to contain only instance methods and properties.
- Removed static methods and properties.
- Updated router.test.ts to align with the changes.
- Replaced calls to Router.setPageHref() and Router.getPageHref() with StaticRouter.setPageHref() and StaticRouter.getPageHref().

Please review these changes and let me know if you have any questions or concerns.

Thank you!

### CODE REVIEW

File: src/router/router.test.ts [hunk 1]
```diff
@@ -1,5 +1,6 @@
 import type { RouterSchema } from '~/types'
 import { Router } from './router'
+import { StaticRouter } from './static-route'
```
Comment (svobik7) [inline src/router/router.test.ts:3]: Can we call this file `static-router`? Instead of `static-route`.

**Review by svobik7 — COMMENTED** [review]
EDIT: see my next comment. Simple inheritance could solve it all https://github.com/svobik7/next-roots/pull/153#issuecomment-1780412938

Hi,

thank you for this nice refactor. Please find the log from vercel bellow. Seems that examples must be adjusted accordingly.

Would you please fix it? Probably even Router.vetPageHref must be updated.

```
TypeError: next_roots__WEBPACK_IMPORTED_MODULE_2__.Router.setPageHref is not a function
    at BlogsAuthorArticlePage (/vercel/path0/examples/basic/.next/server/app/(en)/blogs/[author]/[article]/page.js:341:52)
    at S (/vercel/path0/examples/basic/.next/server/chunks/101.js:6242:13

```

File: src/router/router.ts [hunk 2]
```diff
@@ -1,109 +1,58 @@
...
   }
 
-  /**
-   * Gets relevant page href
-   */
-  public static getPageHref() {
-    return Router.PAGE_HREF
-  }
-
-  /**
```
Comment (svobik7) [inline src/router/router.ts:20]: Can we keep these docs please? At least a comment part. It makes the code visually separated in my eyes.

**Comment by svobik7** [comment]
IMO also templates should be adjusted accordingly. E.g. here https://github.com/svobik7/next-roots/blob/master/src/cli/templates/page-tpl.ts on line 30,33,35 the original Router class is used.

In that case reasonable would be that `Router extends StaticRouter`.

That could solve even the current vercel issue.

What do you think?

File: src/router/router.ts (hunk 2 above)
Comment (adl5423) [reply src/router/router.ts:20]: Hey, I thought those docs were on the code 🥲. Probably erased it accidentally. Will fix it along with the rest of the suggestions.

**Comment by adl5423** [comment]
Incorporated all the fixes but it might be launching the same error again

image

image

File: src/router/router.ts [hunk 3]
```diff
@@ -131,17 +80,10 @@ export function compileHref(
   } catch {
     compiledHref = href
   }
-
   return compiledHref
 }
 
-/**
- * Removes duplicated or trailing slashes from given href and puts the slash at the beginning
```
Comment (svobik7) [inline src/router/router.ts:139]: Please get this comment back

File: src/router/router.ts [hunk 4]
```diff
@@ -112,12 +67,6 @@ export class Router {
   }
 }
 
-/**
- * Puts given params to their appropriate places in given href
```
Comment (svobik7) [inline src/router/router.ts:116]: Please get this comment back

File: src/router/router.ts [hunk 5]
```diff
@@ -13,97 +13,52 @@ export class Router {
...
-   * @returns
-   */
   private findRouteByLocaleAndName(locale: string, name: string) {
     return this.getLocalizedRoutes(locale).find(
       (route: Route) => route.name === name
     )
   }
 
-  /**
-   * Finds route for given locale and href
```
Comment (svobik7) [inline src/router/router.ts:102]: Please get this comment back

File: src/router/router.ts [hunk 6]
```diff
@@ -13,97 +13,52 @@ export class Router {
...
           return dynamicIndexB - dynamicIndexA
         }
-
         return isDynamicA ? 1 : -1
       }) || []
     )
   }
 
-  /**
-   * Finds route for given name and href
```
Comment (svobik7) [inline src/router/router.ts:90]: Please get this comment back

File: src/router/router.ts [hunk 7]
```diff
@@ -13,97 +13,52 @@ export class Router {
...
-   * @param href
-   * @returns
-   */
   public getRouteFromHref(href: string): Route | undefined {
     const locale = this.getLocaleFromHref(href)
     return this.findRouteByLocaleAndHref(locale, href)
   }
 
-  /**
-   * Gets all routes for given locale
```
Comment (svobik7) [inline src/router/router.ts:67]: Please get this comment back

File: src/router/router.ts [hunk 8]
```diff
@@ -13,97 +13,52 @@ export class Router {
...
     const getLocale = getLocaleFactory({
       locales: this.schema.locales,
       defaultLocale: this.schema.defaultLocale,
     })
-
     return getLocale(href)
   }
 
-  /**
-   * Finds route matching given href
```
Comment (svobik7) [inline src/router/router.ts:57]: Please get this comment back

File: src/router/router.ts [hunk 9]
```diff
@@ -13,97 +13,52 @@ export class Router {
...
+      locale = this.getLocaleFromHref(StaticRouter.getPageHref()),
+      ...hrefParams
+    } = params
     const route = this.findRouteByLocaleAndName(locale, name)
-
     return formatHref(compileHref(route?.href || '', hrefParams))
   }
 
-  /**
-   * Retrieves locale of given href
```
Comment (svobik7) [inline src/router/router.ts:43]: Please get this comment back

File: src/router/router.ts [hunk 10]
```diff
@@ -13,97 +13,52 @@ export class Router {
...
-  /**
-   * Sets relevant page href
-   * @param pageHref
-   */
-  public static setPageHref(pageHref: string) {
-    Router.PAGE_HREF = pageHref
-  }
-
-  /**
-   * Creates href by finding route by given name and compiles its href with given params
```
Comment (svobik7) [inline src/router/router.ts:29]: Please get this comment back

File: src/router/router.ts [hunk 11]
```diff
@@ -13,97 +13,52 @@ export class Router {
   /**
    * Gets relevant page href
```
Comment (svobik7) [inline src/router/router.ts:14]: This comment can be removed as the method was moved to StatcRouter

File: src/router/router.ts [hunk 12]
```diff
@@ -1,10 +1,10 @@
 import { compile, match } from 'path-to-regexp'
 import type { Route, RouterSchema } from '~/types'
 import { getLocaleFactory } from '~/utils/locale-utils'
+import { StaticRouter } from './static-router' // Importing separate static router
 
 export class Router {
```
Comment (svobik7) [inline src/router/router.ts:6]: How about `Router extends StaticRouter`? This may fix the issues in examples.

**Review by svobik7 — CHANGES_REQUESTED** [review]
Thank your the update. But it seems incomplete. Could you please address the following before we merge your changes?

- please get the comments back = I marked them all in current review comments
- please fix the issues in examples folder = easiest way might be to extend StaticRouter in Router class

 Vercel logs:
image

**Comment by adl5423** [comment]
Okay, did all the changes: re-incorporated the comments, and improved them. Hopefully it should fix everything now.

**Review by svobik7 — APPROVED** [review]
Good job 👍

**Pull request merged**